Orange/widgets/unsupervised/owkmeans.py

```diff
@@ -172,114 +229,185 @@ def adjustSize(self):
 
     def update_k(self):
         self.optimize_k = False
-        self.apply()
+        self.commit()
 
     def update_from(self):
         self.k_to = max(self.k_from + 1, self.k_to)
         self.optimize_k = True
-        self.apply()
+        self.commit()
 
     def update_to(self):
         self.k_from = min(self.k_from, self.k_to - 1)
         self.optimize_k = True
-        self.apply()
+        self.commit()
 
-    def check_data_size(self, n, msg_group):
-        msg_group.add_message(
-            "not_enough_data",
-            "Too few ({}) unique data instances for {} clusters")
-        data_ok = n <= len(self.data)
-        msg_group.not_enough_data(len(self.data), n, shown=not data_ok)
-        return data_ok
+    def enough_data_instances(self, k):
+        """k cannot be larger than the number of data instances."""
+        return len(self.data) >= k
 
     def _compute_clustering(self, k):
```

Comment:
Please refactor `_compute_clustering` into a pure function with explicit max_iteration, init, n_init, data, ... parameters and change the `__launch_tasks` accordingly.

Reply:
I'm guessing that when we call `executor.submit`, the actual job function isn't called straight away, so the problem with the non-pure code was that the settings may change after applying but before the `_compute_clustering` was actually called. Am I interpreting this correctly?

I've changed it in any case.

Reply:
> Am I interpreting this correctly?

Yes.